cmd/nvidia-dra-plugin/driver.go

```diff
@@ -141,11 +154,30 @@ func (d *driver) nodePrepareResource(ctx context.Context, claim *drapbv1.Claim)
 	return &drapbv1.NodePrepareResourceResponse{CDIDevices: prepared}
 }
 
-func (d *driver) NodeUnprepareResources(ctx context.Context, req *drapbv1.NodeUnprepareResourcesRequest) (*drapbv1.NodeUnprepareResourcesResponse, error) {
-	// We don't upprepare as part of NodeUnprepareResource, we do it
-	// asynchronously when the claims themselves are deleted and the
-	// AllocatedClaim has been removed.
-	return &drapbv1.NodeUnprepareResourcesResponse{}, nil
+func (d *driver) nodeUnprepareResource(ctx context.Context, claim *drapbv1.Claim) *drapbv1.NodeUnprepareResourceResponse {
+	d.Lock()
+	defer d.Unlock()
+
+	isUnprepared, err := d.isUnprepared(ctx, claim.Uid)
+	if err != nil {
+		return &drapbv1.NodeUnprepareResourceResponse{
+			Error: fmt.Sprintf("error checking if claim is already unprepared: %v", err),
+		}
+	}
+
+	if isUnprepared {
+		klog.Infof("Already unprepared, nothing to do for claim '%v'", claim.Uid)
+		return &drapbv1.NodeUnprepareResourceResponse{}
+	}
+
+	err = d.unprepare(ctx, claim.Uid)
+	if err != nil {
+		return &drapbv1.NodeUnprepareResourceResponse{
+			Error: fmt.Sprintf("error unpreparing devices for claim %v: %v", claim.Uid, err),
```

Comment:
Question: Should this be "devices" or "resources"?

Reply:
I actually think either is fine since what we end up returning on success of the prepare calls is a list of CDI devices:
```
&drapbv1.NodePrepareResourceResponse{CDIDevices: prepared}
```